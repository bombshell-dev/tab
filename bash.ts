export function generate(name: string, exec: string) {
}
