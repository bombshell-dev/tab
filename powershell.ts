export function generate(name: string, includeDescription: boolean) {
}
