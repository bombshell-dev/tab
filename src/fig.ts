import type { CommandDef, ArgsDef, PositionalArgDef, CommandMeta } from 'citty';

type FigSpec = {
  name: string;
  description: string;
  options?: FigOption[];
  subcommands?: FigSubcommand[];
  args?: FigArg[];
};

type FigOption = {
  name: string;
  description: string;
  args?: FigArg[];
  isRequired?: boolean;
};

type FigSubcommand = {
  name: string;
  description: string;
  options?: FigOption[];
  subcommands?: FigSubcommand[];
  args?: FigArg[];
};

type FigArg = {
  name: string;
  description?: string;
  isOptional?: boolean;
  isVariadic?: boolean;
  suggestions?: FigSuggestion[];
};

type FigSuggestion = {
  name: string;
  description?: string;
};

async function processArgs<T extends ArgsDef>(
  args: T
): Promise<{ options: FigOption[]; args: FigArg[] }> {
  const options: FigOption[] = [];
  const positionalArgs: FigArg[] = [];

  for (const [name, arg] of Object.entries(args)) {
    if (arg.type === 'positional') {
      const positionalArg = arg as PositionalArgDef;
      positionalArgs.push({
        name,
        description: positionalArg.description,
        isOptional: !positionalArg.required,
        // Assume variadic if the name suggests it (e.g. [...files])
        isVariadic: name.startsWith('[...') || name.startsWith('<...'),
      });
    } else {
      const option: FigOption = {
        name: `--${name}`,
        description: arg.description || '',
        isRequired: arg.required,
      };

      if ('alias' in arg && arg.alias) {
        // Handle both string and array aliases
        const aliases = Array.isArray(arg.alias) ? arg.alias : [arg.alias];
        aliases.forEach((alias) => {
          options.push({
            ...option,
            name: `-${alias}`,
          });
        });
      }

      options.push(option);
    }
  }

  return { options, args: positionalArgs };
}

async function processCommand<T extends ArgsDef>(
  command: CommandDef<T>,
  parentName = ''
): Promise<FigSpec> {
  const resolvedMeta = await Promise.resolve(command.meta);
  const meta = resolvedMeta as CommandMeta;
  const subCommands = await Promise.resolve(command.subCommands);

  if (!meta || !meta.name) {
    throw new Error('Command meta or name is missing');
  }

  const spec: FigSpec = {
    name: parentName ? `${parentName} ${meta.name}` : meta.name,
    description: meta.description || '',
  };

  if (command.args) {
    const resolvedArgs = await Promise.resolve(command.args);
    // Cast to ArgsDef since we know the resolved value will be compatible
    const { options, args } = await processArgs(resolvedArgs as ArgsDef);
    if (options.length > 0) spec.options = options;
    if (args.length > 0) spec.args = args;
  }

  if (subCommands) {
    spec.subcommands = await Promise.all(
      Object.entries(subCommands).map(async ([_, subCmd]) => {
        const resolved = await Promise.resolve(subCmd);
        return processCommand(resolved, spec.name);
      })
    );
  }

  return spec;
}

export async function generateFigSpec<T extends ArgsDef>(
  command: CommandDef<T>
): Promise<string> {
  const spec = await processCommand(command);
  return JSON.stringify(spec, null, 2);
}
