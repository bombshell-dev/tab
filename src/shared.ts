import { Handler } from './index';

export const noopHandler: Handler = () => {
  return [];
};

// TODO (43081j): use type inference some day, so we can type-check
// that the sub commands exist, the options exist, etc.
export interface CompletionConfig {
  handler?: Handler;
  subCommands?: Record<string, CompletionConfig>;
  options?: Record<
    string,
    {
      handler: Handler;
    }
  >;
}

export function assertDoubleDashes(programName: string = 'cli'): void {
  const dashDashIndex = process.argv.indexOf('--');

  if (dashDashIndex === -1) {
    const errorMessage = `Error: You need to use -- to separate completion arguments.\nExample: ${programName} complete -- <args>`;
    console.error(errorMessage);
  }
}
