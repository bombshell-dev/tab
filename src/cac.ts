import * as zsh from './zsh';
import * as bash from './bash';
import * as fish from './fish';
import * as powershell from './powershell';
import type { CAC } from 'cac';
import { Completion } from './';

const execPath = process.execPath;
const processArgs = process.argv.slice(1);
const quotedExecPath = quoteIfNeeded(execPath);
const quotedProcessArgs = processArgs.map(quoteIfNeeded);
const quotedProcessExecArgs = process.execArgv.map(quoteIfNeeded);

const x = `${quotedExecPath} ${quotedProcessExecArgs.join(' ')} ${quotedProcessArgs[0]}`;

function quoteIfNeeded(path: string): string {
  return path.includes(' ') ? `'${path}'` : path;
}

export default function tab(instance: CAC): Completion {
  const completion = new Completion();

  // Add all commands and their options
  for (const cmd of [instance.globalCommand, ...instance.commands]) {
    if (cmd.name === 'complete') continue; // Skip completion command

    // Get positional args info from command usage
    const args = (cmd.rawName.match(/\[.*?\]|\<.*?\>/g) || []).map((arg) =>
      arg.startsWith('[')
    ); // true if optional (wrapped in [])

    // Add command to completion
    const commandName = completion.addCommand(
      cmd.name === '@@global@@' ? '' : cmd.name,
      cmd.description || '',
      args,
      async () => []
    );

    // Add command options
    for (const option of [...instance.globalCommand.options, ...cmd.options]) {
      // Extract short flag from the name if it exists (e.g., "-c, --config" -> "c")
      const shortFlag = option.name.match(/^-([a-zA-Z]), --/)?.[1];
      
      completion.addOption(
        commandName,
        `--${option.name.replace(/^-[a-zA-Z], --/, '')}`, // Remove the short flag part if it exists
        option.description || '',
        async () => [],
        shortFlag
      );
    }
  }

  instance.command('complete [shell]').action(async (shell, extra) => {
    switch (shell) {
      case 'zsh': {
        const script = zsh.generate(instance.name, x);
        console.log(script);
        break;
      }
      case 'bash': {
        const script = bash.generate(instance.name, x);
        console.log(script);
        break;
      }
      case 'fish': {
        const script = fish.generate(instance.name, x);
        console.log(script);
        break;
      }
      case 'powershell': {
        const script = powershell.generate(instance.name, x);
        console.log(script);
        break;
      }
      default: {
        const args: string[] = extra['--'];
        instance.showHelpOnExit = false;

        // Parse current command context
        instance.unsetMatchedCommand();
        instance.parse([execPath, processArgs[0], ...args], {
          run: false,
        });

        // const matchedCommand = instance.matchedCommand?.name || '';
        // const potentialCommand = args.join(' ')
        // console.log(potentialCommand)
        return completion.parse(args);
      }
    }
  });

  return completion;
}
