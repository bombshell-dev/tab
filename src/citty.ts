import { ArgDef, defineCommand } from "citty";
import * as zsh from "../zsh";
import * as bash from "../bash";
import * as fish from "../fish";
import * as powershell from "../powershell";
import { Completion } from ".";
import { CommandDef, PositionalArgDef } from "citty";

function quoteIfNeeded(path) {
  return path.includes(" ") ? `'${path}'` : path;
}

const execPath = process.execPath;
const processArgs = process.argv.slice(1);
const quotedExecPath = quoteIfNeeded(execPath);
const quotedProcessArgs = processArgs.map(quoteIfNeeded);
const quotedProcessExecArgs = process.execArgv.map(quoteIfNeeded);
const x = `${quotedExecPath} ${quotedProcessExecArgs.join(" ")} ${quotedProcessArgs[0]}`;
const completion = new Completion();

export default async function tab(instance: CommandDef) {
  const meta = await resolve(instance.meta);

  if (!meta) {
    throw new Error("Invalid meta.");
  }
  const name = meta.name;
  if (!name) {
    throw new Error("Invalid meta or missing name.");
  }

  const subCommands = await resolve(instance.subCommands);
  if (!subCommands) {
    throw new Error("Invalid or missing subCommands.");
  }

  completion.addCommand(meta.name!, meta?.description ?? "", () => { });

  for (const [cmd, resolvableConfig] of Object.entries(subCommands)) {
    const config = await resolve(resolvableConfig);
    const meta = await resolve(config.meta);

    if (!meta || typeof meta?.description !== "string") {
      throw new Error("Invalid meta or missing description.");
    }

    completion.addCommand(cmd, meta.description, config.run ?? (() => { }));

    if (config.args) {
      for (const [argName, argConfig] of Object.entries(config.args)) {
        const conf = argConfig as ArgDef;
        completion.addOption(
          meta.name!,
          `--${argName}`,
          conf.description ?? "",
          () => { }
        );
      }
    }
  }

  if (instance.args) {
    for (const [argName, argConfig] of Object.entries(instance.args)) {
      const conf = argConfig as PositionalArgDef;
      completion.addOption(
        meta.name!,
        `--${argName}`,
        conf.description ?? "",
        () => { }
      );
    }
  }

  subCommands["complete"] = defineCommand({
    meta: {
      name: "complete",
      description: "Generate shell completion scripts",
    },
    args: {
      shell: {
        type: "positional",
        required: false,
        description: "Specify shell type",
      },
    },
    async run(ctx) {
      let shell: string | undefined = ctx.args.shell;
      if (shell?.startsWith("--")) {
        shell = undefined;
      }

      const extra = ctx.args._ || [];

      switch (shell) {
        case "zsh": {
          const script = zsh.generate(name, x);
          console.log(script);
          break;
        }
        case "bash": {
          const script = bash.generate(name, x);
          console.log(script);
          break;
        }
        case "fish": {
          const script = fish.generate(name, x);
          console.log(script);
          break;
        }
        case "powershell": {
          const script = powershell.generate(name, x);
          console.log(script);
          break;
        }
        default: {
          const extra = ctx.args._ || [];
          await completion.parse(extra, instance);
          return;
        }
      }
    },
  });
}

async function resolve<T>(resolvable: T | (() => T | Promise<T>)): Promise<T> {
  return typeof resolvable === "function" ? await (resolvable as () => T | Promise<T>)() : resolvable;
}