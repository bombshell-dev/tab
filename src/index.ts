import * as zsh from "./zsh";
import * as bash from "./bash";
import * as fish from "./fish";
import * as powershell from "./powershell";

// ShellCompRequestCmd is the name of the hidden command that is used to request
// completion results from the program. It is used by the shell completion scripts.
export const ShellCompRequestCmd: string = "__complete";

// ShellCompNoDescRequestCmd is the name of the hidden command that is used to request
// completion results without their description. It is used by the shell completion scripts.
export const ShellCompNoDescRequestCmd: string = "__completeNoDesc";

// ShellCompDirective is a bit map representing the different behaviors the shell
// can be instructed to have once completions have been provided.
export const ShellCompDirective = {
  // ShellCompDirectiveError indicates an error occurred and completions should be ignored.
  ShellCompDirectiveError: 1 << 0,

  // ShellCompDirectiveNoSpace indicates that the shell should not add a space
  // after the completion even if there is a single completion provided.
  ShellCompDirectiveNoSpace: 1 << 1,

  // ShellCompDirectiveNoFileComp indicates that the shell should not provide
  // file completion even when no completion is provided.
  ShellCompDirectiveNoFileComp: 1 << 2,

  // ShellCompDirectiveFilterFileExt indicates that the provided completions
  // should be used as file extension filters.
  // For flags, using Command.MarkFlagFilename() and Command.MarkPersistentFlagFilename()
  // is a shortcut to using this directive explicitly.  The BashCompFilenameExt
  // annotation can also be used to obtain the same behavior for flags.
  ShellCompDirectiveFilterFileExt: 1 << 3,

  // ShellCompDirectiveFilterDirs indicates that only directory names should
  // be provided in file completion.  To request directory names within another
  // directory, the returned completions should specify the directory within
  // which to search.  The BashCompSubdirsInDir annotation can be used to
  // obtain the same behavior but only for flags.
  ShellCompDirectiveFilterDirs: 1 << 4,

  // ShellCompDirectiveKeepOrder indicates that the shell should preserve the order
  // in which the completions are provided.
  ShellCompDirectiveKeepOrder: 1 << 5,

  // ===========================================================================

  // All directives using iota (or equivalent in Go) should be above this one.
  // For internal use.
  shellCompDirectiveMaxValue: 1 << 6,

  // ShellCompDirectiveDefault indicates to let the shell perform its default
  // behavior after completions have been provided.
  // This one must be last to avoid messing up the iota count.
  ShellCompDirectiveDefault: 0,
};


export type Positional = {
  required: boolean;
  variadic: boolean;
  completion: Handler;
};

type Items = {
    description: string;
    value: string;
}

type Handler = (previousArgs: string[], toComplete: string, endsWithSpace: boolean) => (Items[] | Promise<Items[]>);

type Option = {
    description: string;
    handler: Handler;
}

type Command = {
    description: string;
    handler: Handler;
    options: Map<string, Option>;
    parent?: Command;
}

export class Completion {
    commands = new Map<string, Command>();

    addCommand(name: string, description: string, handler: Handler, parent?: string) {
        const key = parent ? `${parent} ${name}` : name
        this.commands.set(key, { description, handler, options: new Map(), parent: parent ? this.commands.get(parent)! : this.commands.get('')! });
        return key 
    }

    addOption(command: string, option: string, description: string, handler: Handler) {
        const cmd = this.commands.get(command);
        if (!cmd) {
            throw new Error(`Command ${command} not found.`);
        }
        cmd.options.set(option, { description, handler });
        return option
    }

    async parse(args: string[], potentialCommand: string) {
        console.log(potentialCommand)
        const matchedCommand = this.commands.get(potentialCommand) ?? this.commands.get('')!;
        let directive = ShellCompDirective.ShellCompDirectiveDefault;
        const completions: string[] = [];

        const endsWithSpace = args[args.length - 1] === "";
        if (endsWithSpace) {
            args.pop();
        }

        let toComplete = args[args.length - 1] || "";
        const previousArgs = args.slice(0, -1);

        console.log('here', previousArgs)
        if (previousArgs.length > 0) {
            const lastPrevArg = previousArgs[previousArgs.length - 1];
            if (lastPrevArg.startsWith("--") && endsWithSpace) {
                console.log('here')
                const {handler} = matchedCommand.options.get(lastPrevArg)!;
                if (handler) {
                    const flagSuggestions = await handler(previousArgs, toComplete, endsWithSpace);
                    completions.push(
                        ...flagSuggestions.filter(comp => comp.value.startsWith(toComplete)).map(
                            (comp) => `${comp.value}\t${comp.description ?? ""}`
                        )
                    );
                    directive = ShellCompDirective.ShellCompDirectiveNoFileComp;
                    completions.forEach((comp) => console.log(comp));
                    console.log(`:${directive}`);
                    return;
                }
            }
        }

        if (toComplete.startsWith("--")) {
            directive = ShellCompDirective.ShellCompDirectiveNoFileComp;
            const equalsIndex = toComplete.indexOf("=");
            
            if (equalsIndex !== -1) {
                const flagName = toComplete.slice(2, equalsIndex);
                const valueToComplete = toComplete.slice(equalsIndex + 1);
                const {handler} = matchedCommand.options.get(`--${flagName}`)!;
                
                if (handler) {
                    const suggestions = await handler(previousArgs, valueToComplete, endsWithSpace);
                    completions.push(...suggestions.map(
                        (comp) => `${comp.value}\t${comp.description ?? ""}`
                    ));
                }
            } else if (!endsWithSpace) {
                const options = new Map(matchedCommand.options);
                
                let currentCommand = matchedCommand;
                while (currentCommand.parent) {
                    for (const [key, value] of currentCommand.parent.options) {
                        if (!options.has(key)) {
                            options.set(key, value);
                        }
                    }
                    currentCommand = currentCommand.parent;
                }

                const specifiedFlags = previousArgs
                    .filter(arg => arg.startsWith("-"))
                    .filter(arg => arg.startsWith("--"));
                const availableFlags = [...options.keys()]
                    .filter(flag => !specifiedFlags.includes(flag))
                    .filter(flag => flag.startsWith(toComplete));

                completions.push(
                    ...availableFlags.map(
                        (flag) => `${flag}\t${options.get(flag)!.description ?? ""}`
                    )
                );
            } else {
                console.log('here3')
                const {handler} = matchedCommand.options.get(toComplete)!;
                console.log(handler, toComplete)
                
                if (handler) {
                    const suggestions = await handler(previousArgs, toComplete, endsWithSpace);
                    console.log(suggestions)
                    completions.push(...suggestions.map(
                        (comp) => `${comp.value}\t${comp.description ?? ""}`
                    ));
                }

            }
        } else if (!toComplete && endsWithSpace) {
            directive = ShellCompDirective.ShellCompDirectiveNoFileComp;

            completions.push(
                ...Object.keys(this.commands)
                    .filter((cmd) => cmd !== "complete")
                    .map((cmd) => `${cmd}\t${this.commands[cmd].description}`)
            );

            const positionalCompletions = positionalMap.get(potentialCommand) || [];
            for (const positional of positionalCompletions) {
                const suggestions = await positional.completion(previousArgs, "");
                completions.push(
                    ...suggestions.map((comp) => `${comp.action}\t${comp.description ?? ""}`)
                );
                if (suggestions.length > 0) {
                    directive = ShellCompDirective.ShellCompDirectiveNoFileComp;
                }
            }
        }

        completions.forEach((comp) => console.log(comp));
        console.log(`:${directive}`);
    }
}

export function script(shell: "zsh" | "bash" | "fish" | "powershell", name: string, x: string) {
    switch (shell) {
      case "zsh": {
        const script = zsh.generate(name, x);
        console.log(script);
        break;
      }
      case "bash": {
        const script = bash.generate(name, x);
        console.log(script);
        break;
      }
      case "fish": {
        const script = fish.generate(name, x);
        console.log(script);
        break;
      }
      case "powershell": {
        const script = powershell.generate(name, x);
        console.log(script);
        break;
      }
      default: {
        throw new Error(`Unsupported shell: ${shell}`);
      }
    }
}