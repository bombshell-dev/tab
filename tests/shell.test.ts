import { describe, it, expect } from 'vitest';
import * as fish from '../src/fish';
import * as bash from '../src/bash';
import * as zsh from '../src/zsh';
import * as powershell from '../src/powershell';
import { ShellCompDirective } from '../src';

describe('shell completion generators', () => {
  const name = 'testcli';
  const exec = '/usr/bin/node /path/to/testcli';
  const specialName = 'test-cli:app';
  const escapedName = specialName.replace(/[-:]/g, '_');

  describe('fish shell completion', () => {
    it('should generate a valid fish completion script', () => {
      const script = fish.generate(name, exec);

      // Use snapshot testing instead of individual assertions
      expect(script).toMatchSnapshot();
    });

    it('should handle special characters in the name', () => {
      const script = fish.generate(specialName, exec);

      // Use snapshot testing instead of individual assertions
      expect(script).toMatchSnapshot();
    });
  });

  describe('bash shell completion', () => {
    it('should generate a valid bash completion script', () => {
      const script = bash.generate(name, exec);

      // Check that the script contains the shell name
      expect(script).toContain(`# bash completion for ${name}`);

      // Check that the script defines the directives
      expect(script).toContain(
        `readonly ShellCompDirectiveError=${ShellCompDirective.ShellCompDirectiveError}`
      );
      expect(script).toContain(
        `readonly ShellCompDirectiveNoSpace=${ShellCompDirective.ShellCompDirectiveNoSpace}`
      );
      expect(script).toContain(
        `readonly ShellCompDirectiveNoFileComp=${ShellCompDirective.ShellCompDirectiveNoFileComp}`
      );
      expect(script).toContain(
        `readonly ShellCompDirectiveFilterFileExt=${ShellCompDirective.ShellCompDirectiveFilterFileExt}`
      );
      expect(script).toContain(
        `readonly ShellCompDirectiveFilterDirs=${ShellCompDirective.ShellCompDirectiveFilterDirs}`
      );
      expect(script).toContain(
        `readonly ShellCompDirectiveKeepOrder=${ShellCompDirective.ShellCompDirectiveKeepOrder}`
      );

      // Check that the script contains the debug function
      expect(script).toContain(`__${name}_debug()`);

      // Check that the script contains the completion function
      expect(script).toContain(`__${name}_complete()`);

      // Check that the script contains the completion registration
      expect(script).toContain(`complete -F __${name}_complete ${name}`);

      // Check that the script uses the provided exec path
      expect(script).toContain(`requestComp="${exec} complete --`);

      // Check that the script handles directives correctly
      expect(script).toContain(
        `if [[ $((directive & $ShellCompDirectiveError)) -ne 0 ]]`
      );
      expect(script).toContain(
        `if [[ $((directive & $ShellCompDirectiveNoSpace)) -ne 0 ]]`
      );
      expect(script).toContain(
        `if [[ $((directive & $ShellCompDirectiveKeepOrder)) -ne 0 ]]`
      );
      expect(script).toContain(
        `if [[ $((directive & $ShellCompDirectiveNoFileComp)) -ne 0 ]]`
      );
      expect(script).toContain(
        `if [[ $((directive & $ShellCompDirectiveFilterFileExt)) -ne 0 ]]`
      );
      expect(script).toContain(
        `if [[ $((directive & $ShellCompDirectiveFilterDirs)) -ne 0 ]]`
      );
    });

    it('should handle special characters in the name', () => {
      const script = bash.generate(specialName, exec);

      // Check that the script properly escapes the name
      expect(script).toContain(`__${escapedName}_debug()`);
      expect(script).toContain(`__${escapedName}_complete()`);
      expect(script).toContain(
        `complete -F __${escapedName}_complete ${specialName}`
      );
    });
  });

  describe('zsh shell completion', () => {
    it('should generate a valid zsh completion script', () => {
      const script = zsh.generate(name, exec);

      // Check that the script contains the shell name
      expect(script).toContain(`#compdef ${name}`);
      expect(script).toContain(`compdef _${name} ${name}`);

      // Check that the script contains the debug function
      expect(script).toContain(`__${name}_debug()`);

      // Check that the script contains the completion function
      expect(script).toContain(`_${name}()`);

      // Check that the script uses the provided exec path
      expect(script).toContain(`requestComp="${exec} complete --`);

      // Check that the script handles directives
      expect(script).toContain(
        `shellCompDirectiveError=${ShellCompDirective.ShellCompDirectiveError}`
      );
      expect(script).toContain(
        `shellCompDirectiveNoSpace=${ShellCompDirective.ShellCompDirectiveNoSpace}`
      );
      expect(script).toContain(
        `shellCompDirectiveNoFileComp=${ShellCompDirective.ShellCompDirectiveNoFileComp}`
      );
      expect(script).toContain(
        `shellCompDirectiveFilterFileExt=${ShellCompDirective.ShellCompDirectiveFilterFileExt}`
      );
      expect(script).toContain(
        `shellCompDirectiveFilterDirs=${ShellCompDirective.ShellCompDirectiveFilterDirs}`
      );
      expect(script).toContain(
        `shellCompDirectiveKeepOrder=${ShellCompDirective.ShellCompDirectiveKeepOrder}`
      );
    });

    it('should handle special characters in the name', () => {
      const script = zsh.generate(specialName, exec);

      // Check that the script properly escapes the name
      expect(script).toContain(`#compdef ${specialName}`);
      // In zsh, special characters are not escaped in the function name
      expect(script).toContain(`__${specialName}_debug()`);
      expect(script).toContain(`_${specialName}()`);
    });
  });

  describe('powershell completion', () => {
    it('should generate a valid powershell completion script', () => {
      const script = powershell.generate(name, exec);

      // Check that the script contains the shell name
      expect(script).toContain(`# powershell completion for ${name}`);

      // Check that the script contains the debug function
      expect(script).toContain(`function __${name}_debug`);

      // Check that the script contains the completion block
      expect(script).toContain(`[scriptblock]$__${name}CompleterBlock =`);

      // Check that the script uses the provided exec path
      expect(script).toContain(`$RequestComp = "& ${exec} complete --`);

      // Check that the script handles directives
      expect(script).toContain(
        `$ShellCompDirectiveError=${ShellCompDirective.ShellCompDirectiveError}`
      );
      expect(script).toContain(
        `$ShellCompDirectiveNoSpace=${ShellCompDirective.ShellCompDirectiveNoSpace}`
      );
      expect(script).toContain(
        `$ShellCompDirectiveNoFileComp=${ShellCompDirective.ShellCompDirectiveNoFileComp}`
      );
      expect(script).toContain(
        `$ShellCompDirectiveFilterFileExt=${ShellCompDirective.ShellCompDirectiveFilterFileExt}`
      );
      expect(script).toContain(
        `$ShellCompDirectiveFilterDirs=${ShellCompDirective.ShellCompDirectiveFilterDirs}`
      );
      expect(script).toContain(
        `$ShellCompDirectiveKeepOrder=${ShellCompDirective.ShellCompDirectiveKeepOrder}`
      );
    });

    it('should handle special characters in the name', () => {
      const script = powershell.generate(specialName, exec);

      // Check that the script properly escapes the name
      // In PowerShell, special characters are not escaped in the function name
      expect(script).toContain(`function __${specialName}_debug`);
      // The CompleterBlock name uses underscores instead of colons
      expect(script).toContain(`$__test_cli_appCompleterBlock`);
    });
  });
});
