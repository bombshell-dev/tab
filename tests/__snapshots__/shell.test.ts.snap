// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shell completion generators > fish shell completion > should generate a valid fish completion script 1`] = `
"# fish completion for testcli -*- shell-script -*-

# Define shell completion directives
set -l ShellCompDirectiveError 1
set -l ShellCompDirectiveNoSpace 2
set -l ShellCompDirectiveNoFileComp 4
set -l ShellCompDirectiveFilterFileExt 8
set -l ShellCompDirectiveFilterDirs 16
set -l ShellCompDirectiveKeepOrder 32

function __testcli_debug
    set -l file "$BASH_COMP_DEBUG_FILE"
    if test -n "$file"
        echo "$argv" >> $file
    end
end

function __testcli_perform_completion
    __testcli_debug "Starting __testcli_perform_completion"

    # Extract all args except the completion flag
    set -l args (string match -v -- "--completion=" (commandline -opc))
    
    # Extract the current token being completed
    set -l current_token (commandline -ct)
    
    # Check if current token starts with a dash
    set -l flag_prefix ""
    if string match -q -- "-*" $current_token
        set flag_prefix "--flag="
    end
    
    __testcli_debug "Current token: $current_token"
    __testcli_debug "All args: $args"

    # Call the completion program and get the results
    set -l requestComp "/usr/bin/node /path/to/testcli complete -- $args"
    __testcli_debug "Calling $requestComp"
    set -l results (eval $requestComp 2> /dev/null)
    
    # Some programs may output extra empty lines after the directive.
    # Let's ignore them or else it will break completion.
    # Ref: https://github.com/spf13/cobra/issues/1279
    for line in $results[-1..1]
        if test (string sub -s 1 -l 1 -- $line) = ":"
            # The directive
            set -l directive (string sub -s 2 -- $line)
            set -l directive_num (math $directive)
            break
        end
    end
    
    # No directive specified, use default
    if not set -q directive_num
        set directive_num 0
    end
    
    __testcli_debug "Directive: $directive_num"

    # Process completions based on directive
    if test $directive_num -eq $ShellCompDirectiveError
        # Error code. No completion.
        __testcli_debug "Received error directive: aborting."
        return 1
    end

    # Filter out the directive (last line)
    if test (count $results) -gt 0 -a (string sub -s 1 -l 1 -- $results[-1]) = ":"
        set results $results[1..-2]
    end

    # No completions, let fish handle file completions unless forbidden
    if test (count $results) -eq 0
        if test $directive_num -ne $ShellCompDirectiveNoFileComp
            __testcli_debug "No completions, performing file completion"
            return 1
        end
        __testcli_debug "No completions, but file completion forbidden"
        return 0
    end

    # Filter file extensions
    if test $directive_num -eq $ShellCompDirectiveFilterFileExt
        __testcli_debug "File extension filtering"
        set -l file_extensions
        for item in $results
            if test -n "$item" -a (string sub -s 1 -l 1 -- $item) != "-"
                set -a file_extensions "*$item"
            end
        end
        __testcli_debug "File extensions: $file_extensions"
        
        # Use the file extensions as completions
        set -l completions
        for ext in $file_extensions
            # Get all files matching the extension
            set -a completions (string replace -r '^.*/' '' -- $ext)
        end
        
        for item in $completions
            echo -e "$item	"
        end
        return 0
    end

    # Filter directories
    if test $directive_num -eq $ShellCompDirectiveFilterDirs
        __testcli_debug "Directory filtering"
        set -l dirs
        for item in $results
            if test -d "$item"
                set -a dirs "$item/"
            end
        end
        
        for item in $dirs
            echo -e "$item	"
        end
        return 0
    end

    # Process remaining completions
    for item in $results
        if test -n "$item"
            # Check if the item has a description
            if string match -q "*	*" -- "$item"
                set -l completion_parts (string split 	 -- "$item")
                set -l comp $completion_parts[1]
                set -l desc $completion_parts[2]
                
                # Add the completion and description
                echo -e "$comp	$desc"
            else
                # Add just the completion
                echo -e "$item	"
            end
        end
    end
    
    # If directive contains NoSpace, tell fish not to add a space after completion
    if test (math "$directive_num & $ShellCompDirectiveNoSpace") -ne 0
        return 2
    end
    
    return 0
end

# Set up the completion for the testcli command
complete -c testcli -f -a "(eval __testcli_perform_completion)"
"
`;

exports[`shell completion generators > fish shell completion > should handle special characters in the name 1`] = `
"# fish completion for test-cli:app -*- shell-script -*-

# Define shell completion directives
set -l ShellCompDirectiveError 1
set -l ShellCompDirectiveNoSpace 2
set -l ShellCompDirectiveNoFileComp 4
set -l ShellCompDirectiveFilterFileExt 8
set -l ShellCompDirectiveFilterDirs 16
set -l ShellCompDirectiveKeepOrder 32

function __test_cli_app_debug
    set -l file "$BASH_COMP_DEBUG_FILE"
    if test -n "$file"
        echo "$argv" >> $file
    end
end

function __test_cli_app_perform_completion
    __test_cli_app_debug "Starting __test_cli_app_perform_completion"

    # Extract all args except the completion flag
    set -l args (string match -v -- "--completion=" (commandline -opc))
    
    # Extract the current token being completed
    set -l current_token (commandline -ct)
    
    # Check if current token starts with a dash
    set -l flag_prefix ""
    if string match -q -- "-*" $current_token
        set flag_prefix "--flag="
    end
    
    __test_cli_app_debug "Current token: $current_token"
    __test_cli_app_debug "All args: $args"

    # Call the completion program and get the results
    set -l requestComp "/usr/bin/node /path/to/testcli complete -- $args"
    __test_cli_app_debug "Calling $requestComp"
    set -l results (eval $requestComp 2> /dev/null)
    
    # Some programs may output extra empty lines after the directive.
    # Let's ignore them or else it will break completion.
    # Ref: https://github.com/spf13/cobra/issues/1279
    for line in $results[-1..1]
        if test (string sub -s 1 -l 1 -- $line) = ":"
            # The directive
            set -l directive (string sub -s 2 -- $line)
            set -l directive_num (math $directive)
            break
        end
    end
    
    # No directive specified, use default
    if not set -q directive_num
        set directive_num 0
    end
    
    __test_cli_app_debug "Directive: $directive_num"

    # Process completions based on directive
    if test $directive_num -eq $ShellCompDirectiveError
        # Error code. No completion.
        __test_cli_app_debug "Received error directive: aborting."
        return 1
    end

    # Filter out the directive (last line)
    if test (count $results) -gt 0 -a (string sub -s 1 -l 1 -- $results[-1]) = ":"
        set results $results[1..-2]
    end

    # No completions, let fish handle file completions unless forbidden
    if test (count $results) -eq 0
        if test $directive_num -ne $ShellCompDirectiveNoFileComp
            __test_cli_app_debug "No completions, performing file completion"
            return 1
        end
        __test_cli_app_debug "No completions, but file completion forbidden"
        return 0
    end

    # Filter file extensions
    if test $directive_num -eq $ShellCompDirectiveFilterFileExt
        __test_cli_app_debug "File extension filtering"
        set -l file_extensions
        for item in $results
            if test -n "$item" -a (string sub -s 1 -l 1 -- $item) != "-"
                set -a file_extensions "*$item"
            end
        end
        __test_cli_app_debug "File extensions: $file_extensions"
        
        # Use the file extensions as completions
        set -l completions
        for ext in $file_extensions
            # Get all files matching the extension
            set -a completions (string replace -r '^.*/' '' -- $ext)
        end
        
        for item in $completions
            echo -e "$item	"
        end
        return 0
    end

    # Filter directories
    if test $directive_num -eq $ShellCompDirectiveFilterDirs
        __test_cli_app_debug "Directory filtering"
        set -l dirs
        for item in $results
            if test -d "$item"
                set -a dirs "$item/"
            end
        end
        
        for item in $dirs
            echo -e "$item	"
        end
        return 0
    end

    # Process remaining completions
    for item in $results
        if test -n "$item"
            # Check if the item has a description
            if string match -q "*	*" -- "$item"
                set -l completion_parts (string split 	 -- "$item")
                set -l comp $completion_parts[1]
                set -l desc $completion_parts[2]
                
                # Add the completion and description
                echo -e "$comp	$desc"
            else
                # Add just the completion
                echo -e "$item	"
            end
        end
    end
    
    # If directive contains NoSpace, tell fish not to add a space after completion
    if test (math "$directive_num & $ShellCompDirectiveNoSpace") -ne 0
        return 2
    end
    
    return 0
end

# Set up the completion for the test-cli:app command
complete -c test-cli:app -f -a "(eval __test_cli_app_perform_completion)"
"
`;
